plugins/node/opentelemetry-instrumentation-bunyan/src/bunyan.ts

```diff
@@ -0,0 +1,101 @@
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { context, getSpan, isSpanContextValid } from '@opentelemetry/api';
+import {
+  InstrumentationBase,
+  InstrumentationNodeModuleDefinition,
+  isWrapped,
+} from '@opentelemetry/instrumentation';
+import { BunyanInstrumentationConfig } from './types';
+import { VERSION } from './version';
+import type * as BunyanLogger from 'bunyan';
+
+function defaultHook() {}
+
+export class BunyanInstrumentation extends InstrumentationBase<
+  typeof BunyanLogger
+> {
+  constructor(config: BunyanInstrumentationConfig = { logHook: defaultHook }) {
+    super('@opentelemetry/instrumentation-bunyan', VERSION, config);
+  }
+
+  protected init() {
+    return [
+      new InstrumentationNodeModuleDefinition<typeof BunyanLogger>(
+        'bunyan',
+        ['*'],
+        logger => {
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          const proto = logger.prototype as any;
+          if (isWrapped(proto['_emit'])) {
+            this._unwrap(proto, '_emit');
+          }
+
+          this._wrap(
+            proto,
+            '_emit',
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+            this._getPatchedEmit() as any
+          );
+          return logger;
+        },
+        logger => {
+          if (logger === undefined) return;
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          this._unwrap(logger.prototype as any, '_emit');
+        }
+      ),
+    ];
+  }
+
+  getConfig(): BunyanInstrumentationConfig {
+    return this._config;
+  }
+
+  setConfig(config: BunyanInstrumentationConfig) {
+    this._config = Object.assign({ logHook: defaultHook }, config);
+  }
+
+  private _getPatchedEmit() {
+    return (original: (...args: unknown[]) => void) => {
+      const instrumentation = this;
+      return function patchedEmit(this: BunyanLogger, ...args: unknown[]) {
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        const record = args[0] as any;
+        const span = getSpan(context.active());
+
+        if (!span) {
+          return original.apply(this, args);
+        }
+
+        const spanContext = span.context();
+
+        if (!isSpanContextValid(spanContext)) {
+          return original.apply(this, args);
+        }
+
+        record['trace_id'] = spanContext.traceId;
+        record['span_id'] = spanContext.spanId;
+        record['trace_flags'] = `0${spanContext.traceFlags.toString(16)}`;
+
+        instrumentation.getConfig().logHook!(record, span);
```

Comment:
It is usually common to surround hook functions with `safeExecuteInTheMiddle`, so if the hook function throws, the exception will be catched, ignored, and app will not crash.

Reply:
Although is it good to have silent failures? Of course the instrumentation should not throw anything, but if it is the user code and user code throws, then I'm not so sure :thinking:

Reply:
I'm not against it if it is the convention here, but is it The Right Thing :tm:? :smile:

Reply:
In ioredis instrumentation I was asked in the review to log an error in case **hook function** throws:
```js
        safeExecuteInTheMiddle(
          () => config?.responseHook?.(span, cmd.name, cmd.args, result),
          e => {
            if (e) {
              diag.error('ioredis response hook failed', e);
            }
          },
          true
        );
```

As far as I understand, If **`original.apply(this, arguments)`** throws, then `safeExecuteInTheMiddle` should have the parameter `preventThrowingError` set to false, so the exception can propagate back to the caller, and the `onFinish` function is used to close spans or set them to error.

Reply:
I guess, that would be better indeed, thanks :slightly_smiling_face: Adding a fix